File: iosApp/iosApp/Views/Sheets/SeerrRequestSheet.swift
//
//  SeerrRequestSheet.swift
//  iosApp
//

import SwiftUI
import Shared

struct SeerrRequestSheet: View {
    let details: RequestMediaDetails
    let serviceDetails: ServiceDetails?
    let currentUser: SeerrUser?
    let users: [SeerrUser]
    let onDismiss: () -> Void
    let onSubmit: (Int64?, String?, Int64?, [KotlinInt]?, Int64?) -> Void
    
    @State private var selectedProfileId: Int64?
    @State private var selectedRootFolder: String?
    @State private var selectedUserId: Int64?
    @State private var selectedSeasons: Set<Int32> = []
    
    init(
        details: RequestMediaDetails,
        serviceDetails: ServiceDetails?,
        currentUser: SeerrUser?,
        users: [SeerrUser],
        onDismiss: @escaping () -> Void,
        onSubmit: @escaping (Int64?, String?, Int64?, [KotlinInt]?, Int64?) -> Void
    ) {
        self.details = details
        self.serviceDetails = serviceDetails
        self.currentUser = currentUser
        self.users = users
        self.onDismiss = onDismiss
        self.onSubmit = onSubmit
        
        // Initialize state with default values from service details
        _selectedProfileId = State(initialValue: serviceDetails.map { Int64($0.server.activeProfileId) })
        _selectedRootFolder = State(initialValue: serviceDetails?.server.activeDirectory)
        _selectedUserId = State(initialValue: currentUser?.id)
        
        if let tvDetails = details as? TvDetails {
            _selectedSeasons = State(initialValue: Set(tvDetails.seasons.map { $0.seasonNumber }))
        }
    }
    
    var body: some View {
        NavigationView {
            Form {
                Section {
                    HStack(spacing: 16) {
                        GenericPosterItem(posterUrl: details.fullPosterPath)
                            .frame(width: 80)
                        
                        VStack(alignment: .leading, spacing: 4) {
                            Text(details is TvDetails ? MR.strings().type_series.localized().uppercased() : MR.strings().type_movie.localized().uppercased())
                                .font(.caption.bold())
                                .foregroundColor(.accentColor)
                            
                            Text(details.displayTitle)
                                .font(.headline)
                                .lineLimit(2)
                        }
                    }
                    .padding(.vertical, 8)
                }
                
                if let tvDetails = details as? TvDetails {
                    Section(header: Text(MR.strings().seasons_header.localized())) {
                        Toggle(MR.strings().all_seasons.localized(), isOn: Binding(
                            get: { selectedSeasons.count == tvDetails.seasons.count },
                            set: { isOn in
                                if isOn {
                                    selectedSeasons = Set(tvDetails.seasons.map { $0.seasonNumber })
                                } else {
                                    selectedSeasons = []
                                }
                            }
                        ))
                        
                        ForEach(tvDetails.seasons, id: \.seasonNumber) { season in
                            Toggle(isOn: Binding(
                                get: { selectedSeasons.contains(season.seasonNumber) },
                                set: { isOn in
                                    if isOn {
                                        selectedSeasons.insert(season.seasonNumber)
                                    } else {
                                        selectedSeasons.remove(season.seasonNumber)
                                    }
                                }
                            )) {
                                VStack(alignment: .leading) {
                                    Text(season.seasonNumber == 0 ? MR.strings().specials.localized() : MR.strings().season_label.formatted(args: [season.seasonNumber]))
                                    Text(MR.plurals().episodes.localized(season.episodeCount))
                                        .font(.caption)
                                        .foregroundColor(.secondary)
                                }
                            }
                        }
                    }
                }
                
                Section(header: Text(MR.strings().advanced.localized())) {
                    Picker(MR.strings().quality_profile.localized(), selection: $selectedProfileId) {
                        Text(MR.strings().unknown.localized()).tag(nil as Int64?)
                        ForEach(serviceDetails?.profiles ?? [], id: \.id) { profile in
                            Text(profile.name).tag(profile.id as Int64?)
                        }
                    }
                    
                    Picker(MR.strings().root_folder.localized(), selection: $selectedRootFolder) {
                        Text(MR.strings().unknown.localized()).tag(nil as String?)
                        ForEach(serviceDetails?.rootFolders ?? [], id: \.path) { folder in
                            Text(folder.path).tag(folder.path as String?)
                        }
                    }
                    
                    if currentUser?.hasPermission(permission: .admin) == true && !users.isEmpty {
                        Picker(MR.strings().request_as.localized(), selection: $selectedUserId) {
                            ForEach(users, id: \.id) { user in
                                Text(user.displayName).tag(user.id as Int64?)
                            }
                        }
                    }
                }
            }
            .navigationTitle(MR.strings().request.localized())
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarLeading) {
                    Button(MR.strings().cancel.localized()) {
                        onDismiss()
                    }
                }
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button(MR.strings().request.localized()) {
                        let seasons = details is TvDetails ? selectedSeasons.map { KotlinInt(value: $0) } : nil
                        onSubmit(selectedProfileId, selectedRootFolder, nil, seasons, selectedUserId)
                    }
                    .disabled(details is TvDetails && selectedSeasons.isEmpty)
                }
            }
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/dnfapps/arrmatey/ui/sheets/SeerrRequestSheet.kt
package com.dnfapps.arrmatey.ui.sheets

import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.dnfapps.arrmatey.seerr.api.model.RequestMediaDetails
import com.dnfapps.arrmatey.seerr.api.model.SeerrUser
import com.dnfapps.arrmatey.seerr.api.model.ServiceDetails
import com.dnfapps.arrmatey.seerr.api.model.TvDetails
import com.dnfapps.arrmatey.seerr.api.model.UserPermission
import com.dnfapps.arrmatey.shared.MR
import com.dnfapps.arrmatey.ui.components.ContainerCard
import com.dnfapps.arrmatey.ui.components.DropdownPicker
import com.dnfapps.arrmatey.ui.components.LabelledSwitch
import com.dnfapps.arrmatey.utils.mokoPlural
import com.dnfapps.arrmatey.utils.mokoString

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SeerrRequestSheet(
    details: RequestMediaDetails,
    serviceDetails: ServiceDetails?,
    currentUser: SeerrUser?,
    users: List<SeerrUser>,
    requestInProgress: Boolean,
    onDismissRequest: () -> Unit,
    onSubmitRequest: (Long?, String?, Long?, List<Int>?, Long?) -> Unit,
    canSwitchToAddDirectly: Boolean = false,
    instanceTypeName: String? = null,
    onSwitchToAddDirectly: (() -> Unit)? = null,
    canRequest4k: Boolean = false,
    is4k: Boolean = false,
    onIs4kChange: (Boolean) -> Unit = {},
) {
    ModalBottomSheet(
        onDismissRequest = {
            if (!requestInProgress) {
                onDismissRequest()
            }
        },
        sheetState =
            rememberModalBottomSheetState(
                skipPartiallyExpanded = true,
                confirmValueChange = { !requestInProgress },
            ),
    ) {
        SeerrRequestSheetContent(
            details = details,
            serviceDetails = serviceDetails,
            currentUser = currentUser,
            users = users,
            requestInProgress = requestInProgress,
            onDismissRequest = onDismissRequest,
            onSubmitRequest = onSubmitRequest,
            canSwitchToAddDirectly = canSwitchToAddDirectly,
            instanceTypeName = instanceTypeName,
            onSwitchToAddDirectly = onSwitchToAddDirectly,
            canRequest4k = canRequest4k,
            is4k = is4k,
            onIs4kChange = onIs4kChange,
        )
    }
}

@Composable
fun SeerrRequestSheetContent(
    details: RequestMediaDetails,
    serviceDetails: ServiceDetails?,
    currentUser: SeerrUser?,
    users: List<SeerrUser>,
    requestInProgress: Boolean,
    onDismissRequest: () -> Unit,
    onSubmitRequest: (Long?, String?, Long?, List<Int>?, Long?) -> Unit,
    canSwitchToAddDirectly: Boolean = false,
    instanceTypeName: String? = null,
    onSwitchToAddDirectly: (() -> Unit)? = null,
    canRequest4k: Boolean = false,
    is4k: Boolean = false,
    onIs4kChange: (Boolean) -> Unit = {},
    modifier: Modifier = Modifier,
) {
    var selectedProfileId by remember { mutableStateOf<Long?>(null) }
    var selectedRootFolder by remember { mutableStateOf<String?>(null) }
    var selectedUserId by remember { mutableStateOf<Long?>(null) }

    LaunchedEffect(serviceDetails, currentUser) {
        if (selectedProfileId == null) {
            selectedProfileId = serviceDetails?.server?.activeProfileId?.toLong()
        }
        if (selectedRootFolder == null) {
            selectedRootFolder = serviceDetails?.server?.activeDirectory
        }
        if (selectedUserId == null) {
            selectedUserId = currentUser?.id
        }
    }

    var selectedSeasons by remember {
        mutableStateOf(
            if (details is TvDetails) {
                details.seasons.map { it.seasonNumber }.toSet()
            } else {
                emptySet()
            },
        )
    }

    val isAdmin = currentUser?.hasPermission(UserPermission.ADMIN) == true

    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .padding(horizontal = 24.dp)
                .padding(bottom = 24.dp)
                .animateContentSize()
                .verticalScroll(rememberScrollState()),
        verticalArrangement = Arrangement.spacedBy(24.dp),
    ) {
        Column {
            Text(
                text =
                    mokoString(
                        if (details is TvDetails) MR.strings.type_series else MR.strings.type_movie,
                    ).uppercase(),
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.primary,
                fontWeight = FontWeight.Bold,
            )
            Text(
                text = details.displayTitle,
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.Bold,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
        }

        if (canSwitchToAddDirectly && onSwitchToAddDirectly != null) {
            SingleChoiceSegmentedButtonRow(
                modifier = Modifier.fillMaxWidth(),
            ) {
                SegmentedButton(
                    selected = true,
                    onClick = {},
                    shape = SegmentedButtonDefaults.itemShape(index = 0, count = 2),
                    label = { Text(mokoString(MR.strings.request)) },
                )
                SegmentedButton(
                    selected = false,
                    onClick = onSwitchToAddDirectly,
                    shape = SegmentedButtonDefaults.itemShape(index = 1, count = 2),
                    label = { Text(mokoString(MR.strings.add_to_arr, instanceTypeName ?: "Arr")) },
                )
            }
        }

        SeerrRequestConfigurationContent(
            details = details,
            serviceDetails = serviceDetails,
            isAdmin = isAdmin,
            users = users,
            selectedProfileId = selectedProfileId,
            onSelectedProfileIdChange = { selectedProfileId = it },
            selectedRootFolder = selectedRootFolder,
            onSelectedRootFolderChange = { selectedRootFolder = it },
            selectedUserId = selectedUserId,
            onSelectedUserIdChange = { selectedUserId = it },
            selectedSeasons = selectedSeasons,
            onSelectedSeasonsChange = { selectedSeasons = it },
            canRequest4k = canRequest4k,
            is4k = is4k,
            onIs4kChange = onIs4kChange,
            enabled = !requestInProgress,
        )

        Row(
            modifier = Modifier.fillMaxWidth().padding(top = 12.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            OutlinedButton(
                onClick = onDismissRequest,
                modifier = Modifier.weight(1f),
                enabled = !requestInProgress,
            ) {
                Text(mokoString(MR.strings.cancel))
            }
            Button(
                onClick = {
                    val seasons = if (details is TvDetails) selectedSeasons.toList() else null
                    onSubmitRequest(
                        selectedProfileId,
                        selectedRootFolder,
                        null, // languageProfileId
                        seasons,
                        selectedUserId,
                    )
                },
                modifier = Modifier.weight(1f),
                enabled = !requestInProgress && if (details is TvDetails) selectedSeasons.isNotEmpty() else true,
            ) {
                if (requestInProgress) {
                    CircularProgressIndicator(Modifier.size(24.dp))
                } else {
                    Text(
                        if (is4k) {
                            mokoString(MR.strings.request_in_4k)
                        } else {
                            mokoString(MR.strings.request)
                        },
                    )
                }
            }
        }
    }
}

@Composable
fun SeerrRequestConfigurationContent(
    details: RequestMediaDetails,
    serviceDetails: ServiceDetails?,
    isAdmin: Boolean,
    users: List<SeerrUser>,
    selectedProfileId: Long?,
    onSelectedProfileIdChange: (Long?) -> Unit,
    selectedRootFolder: String?,
    onSelectedRootFolderChange: (String?) -> Unit,
    selectedUserId: Long?,
    onSelectedUserIdChange: (Long?) -> Unit,
    selectedSeasons: Set<Int>,
    onSelectedSeasonsChange: (Set<Int>) -> Unit,
    canRequest4k: Boolean = false,
    is4k: Boolean = false,
    onIs4kChange: (Boolean) -> Unit = {},
    enabled: Boolean = true,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        if (canRequest4k) {
            LabelledSwitch(
                label = mokoString(MR.strings.request_in_4k),
                checked = is4k,
                onCheckedChange = onIs4kChange,
                enabled = enabled,
            )
        }

        if (details is TvDetails) {
            SeasonSelector(
                details = details,
                selectedSeasons = selectedSeasons,
                onSeasonsChanged = onSelectedSeasonsChange,
                enabled = enabled,
            )
        }

        Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
            Text(
                text = mokoString(MR.strings.advanced).uppercase(),
                style = MaterialTheme.typography.labelLarge,
                fontWeight = FontWeight.Bold,
            )

            val profiles = serviceDetails?.profiles ?: emptyList()
            val rootFolders = serviceDetails?.rootFolders ?: emptyList()

            DropdownPicker(
                label = { Text(mokoString(MR.strings.quality_profile)) },
                options = profiles.map { it.id }.ifEmpty { listOfNotNull(selectedProfileId) },
                selectedOption = selectedProfileId,
                onOptionSelected = onSelectedProfileIdChange,
                getOptionLabel = { profileId ->
                    profiles.find { it.id == profileId }?.name ?: profileId.toString()
                },
                enabled = enabled,
            )

            DropdownPicker(
                label = { Text(mokoString(MR.strings.root_folder)) },
                options = rootFolders.map { it.path }.ifEmpty { listOfNotNull(selectedRootFolder) },
                selectedOption = selectedRootFolder,
                onOptionSelected = onSelectedRootFolderChange,
                getOptionLabel = { it },
                enabled = enabled,
            )

            if (isAdmin && users.isNotEmpty()) {
                DropdownPicker(
                    label = { Text(mokoString(MR.strings.request_as)) },
                    options = users.map { it.id },
                    selectedOption = selectedUserId,
                    onOptionSelected = onSelectedUserIdChange,
                    getOptionLabel = { userId ->
                        users.find { it.id == userId }?.displayName ?: userId.toString()
                    },
                    enabled = enabled,
                )
            }
        }
    }
}

@Composable
private fun SeasonSelector(
    details: TvDetails,
    selectedSeasons: Set<Int>,
    onSeasonsChanged: (Set<Int>) -> Unit,
    enabled: Boolean = true,
) {
    ContainerCard(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        val allSeasonsSelected = details.seasons.all { it.seasonNumber in selectedSeasons }

        LabelledSwitch(
            label = mokoString(MR.strings.all_seasons),
            checked = allSeasonsSelected,
            onCheckedChange = { checked ->
                if (checked) {
                    onSeasonsChanged(details.seasons.map { it.seasonNumber }.toSet())
                } else {
                    onSeasonsChanged(emptySet())
                }
            },
            enabled = enabled,
        )
        HorizontalDivider()

        details.seasons.forEach { season ->
            val isSelected = season.seasonNumber in selectedSeasons

            LabelledSwitch(
                label =
                    if (season.seasonNumber == 0) {
                        mokoString(MR.strings.specials)
                    } else {
                        mokoString(MR.strings.season_label, season.seasonNumber)
                    },
                sublabel = mokoPlural(MR.plurals.episodes, season.episodeCount),
                checked = isSelected,
                onCheckedChange = { checked ->
                    if (checked) {
                        onSeasonsChanged(selectedSeasons + season.seasonNumber)
                    } else {
                        onSeasonsChanged(selectedSeasons - season.seasonNumber)
                    }
                },
                enabled = enabled,
            )
        }
    }
}
